src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,319 @@
+                int usedBlockCount = _blockCount - firstUsedBlock;
+                if (usedBlockCount + newBlocksNeeded <= _blocks.Length)
+                {
+                    // We can shift the array down to make enough space
+                    _blocks.AsSpan().Slice(firstUsedBlock, usedBlockCount).CopyTo(_blocks);
+                }
+                else
+                {
+                    // Need to reallocate the array
+                    int blockArraySize = _blocks.Length;
+                    while (blockArraySize < usedBlockCount + newBlocksNeeded)
+                    {
+                        blockArraySize *= 2;
+                    }
+
+                    byte[]?[] newBlockArray = new byte[]?[blockArraySize];
+                    _blocks.AsSpan().Slice(firstUsedBlock, usedBlockCount).CopyTo(newBlockArray);
+                    _blocks = newBlockArray;
+                }
+
+                _blockCount = usedBlockCount;
+                _activeStart -= firstUsedBlock * BlockSize;
+                _availableStart -= firstUsedBlock * BlockSize;
+
+                Debug.Assert(_activeStart / BlockSize == 0, $"Start is not in first block after move or resize?? _activeStart={_activeStart}");
+            }
+
+            Debug.Assert(_blockCount + newBlocksNeeded <= _blocks.Length, $"Not enough room for new blocks?? _blockCount={_blockCount}, newBlocksNeeded={newBlocksNeeded}, _blocks.Length={_blocks.Length}");
+
+            // Allocate new blocks
+            for (int i = 0; i < newBlocksNeeded; i++)
+            {
+                _blocks[_blockCount + i] = ArrayPool<byte>.Shared.Rent(BlockSize);
+            }
+
+            _blockCount += newBlocksNeeded;
+
+            Debug.Assert(byteCount <= AvailableMemory.Length);
+        }
+    }
+
+    // This is a Memory-like struct for handling multi-array segments from MultiArrayBuffer above.
+    // It supports standard Span/Memory operations like indexing, Slice, Length, etc
+    // It also supports CopyTo/CopyFrom Span<byte>
+
+    internal readonly struct MultiMemory
+    {
+        private readonly byte[][] _blocks;
+        private readonly int _start;
+        private readonly int _length;
+
+        private const int BlockSize = 16 * 1024;
+
+        internal MultiMemory(byte[][] blocks, int start, int length)
+        {
+            if (length == 0)
+            {
+                _blocks = Array.Empty<byte[]>();
+                _start = 0;
+                _length = 0;
+            }
+            else
+            {
+                Debug.Assert(start >= 0);
+                Debug.Assert(length >= 0);
+                Debug.Assert(start + length <= blocks.Length * BlockSize);
+
+                _blocks = blocks;
+                _start = start;
+                _length = length;
+            }
+        }
+
+        private static int GetBlockIndex(int offset) => offset / BlockSize;
+        private static int GetOffsetInBlock(int offset) => offset % BlockSize;
```

Comment:
Cast to `uint` produces the bit-hack modulo by using `and` instead of the optimized division.
```suggestion
        private static int GetOffsetInBlock(int offset) => (int)((uint)offset % BlockSize);
```